### PR TITLE
Bug: Dirichlet distribution got invalid concentration parameter for BetaDistribution

### LINKED ISSUE #1206: Bug: Dirichlet distribution got invalid concentration parameter for BetaDistribution
I'm still working on creating a minimum viable code to reproduce that doesn't require sharing my data, but when using the BetaProportion distribution I've obtained the below error:

```
ValueError                                Traceback (most recent call last)
<ipython-input-51-4f5c4ea19e62> in <module>()
     22     progress_bar=True,
     23   )
---> 24   mcmc.run(rng_key, 7, reproduced, year_published, code_available, theory, empirical, tfpg, types, topic, citation_count)
     25 

/usr/local/lib/python3.7/dist-packages/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    570         map_args = (rng_key, init_state, init_params)
    571         if self.num_chains == 1:
--> 572             states_flat, last_state = partial_map_fn(map_args)
    573             states = tree_map(lambda x: x[jnp.newaxis, ...], states_flat)
    574         else:

/usr/local/lib/python3.7/dist-packages/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    363                 init_params,
    364                 model_args=args,
--> 365                 model_kwargs=kwargs,
    366             )
    367         sample_fn, postprocess_fn = self._get_cached_fns()

/usr/local/lib/python3.7/dist-packages/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    695             )
    696         init_params = self._init_state(
--> 697             rng_key_init_model, model_args, model_kwargs, init_params
    698         )
    699         if self._potential_fn and init_params is None:

/usr/local/lib/python3.7/dist-packages/numpyro/infer/hmc.py in _init_state(self, rng_key, model_args, model_kwargs, init_params)
    647                 model_args=model_args,
    648                 model_kwargs=model_kwargs,
--> 649                 forward_mode_differentiation=self._forward_mode_differentiation,
    650             )
    651             if self._init_fn is None:

/usr/local/lib/python3.7/dist-packages/numpyro/infer/util.py in initialize_model(rng_key, model, init_strategy, dynamic_args, model_args, model_kwargs, forward_mode_differentiation, validate_grad)
    639             with numpyro.validation_enabled(), trace() as tr:
    640                 # validate parameters
--> 641                 substituted_model(*model_args, **model_kwargs)
    642                 # validate values
    643                 for site in tr.values():

/usr/local/lib/python3.7/dist-packages/numpyro/primitives.py in __call__(self, *args, **kwargs)
     85             return self
     86         with self:
---> 87             return self.fn(*args, **kwargs)
     88 
     89 

/usr/local/lib/python3.7/dist-packages/numpyro/primitives.py in __call__(self, *args, **kwargs)
     85             return self
     86         with self:
---> 87             return self.fn(*args, **kwargs)
     88 
     89 

<ipython-input-49-e65da1086e6f> in model(max_clusters, reproduced, year_published, code_available, theory, empirical, tfpg, types, topic, citation_count)
     19   with numpyro.plate('All Years', max_year-min_year):
     20     with numpyro.plate('Topic Area', n_topics):# shape: (n_topics, max_year-min_year)
---> 21       topic_year_m = sample("Topic-Year Influence", dist.BetaProportion(mean=1, concentration=1))
     22 
     23   print(topic_year_m.shape)

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/distribution.py in __call__(cls, *args, **kwargs)
     92             if result is not None:
     93                 return result
---> 94         return super().__call__(*args, **kwargs)
     95 
     96     @property

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/continuous.py in __init__(self, mean, concentration, validate_args)
   1579             mean * concentration,
   1580             (1.0 - mean) * concentration,
-> 1581             validate_args=validate_args,
   1582         )

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/continuous.py in __init__(self, concentration1, concentration0, validate_args)
     74         concentration0 = jnp.broadcast_to(concentration0, batch_shape)
     75         self._dirichlet = Dirichlet(
---> 76             jnp.stack([concentration1, concentration0], axis=-1)
     77         )
     78         super(Beta, self).__init__(batch_shape=batch_shape, validate_args=validate_args)

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/distribution.py in __call__(cls, *args, **kwargs)
     92             if result is not None:
     93                 return result
---> 94         return super().__call__(*args, **kwargs)
     95 
     96     @property

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/continuous.py in __init__(self, concentration, validate_args)
    157             batch_shape=batch_shape,
    158             event_shape=event_shape,
--> 159             validate_args=validate_args,
    160         )
    161 

/usr/local/lib/python3.7/dist-packages/numpyro/distributions/distribution.py in __init__(self, batch_shape, event_shape, validate_args)
    176                         raise ValueError(
    177                             "{} distribution got invalid {} parameter.".format(
--> 178                                 self.__class__.__name__, param
    179                             )
    180                         )

ValueError: Dirichlet distribution got invalid concentration parameter.
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #1206 

Couple of changes needed to happen:
- the `mean` parameter of `BetaProportion` should be between 0 and 1 not inclusive of the extremes, otherwise one of the `concentration` parameters will be `0`
- the initialisation of `Beta` needs to happen before that of `Dirichelet`, as suggested in the issue

### CODE REVIEW

File: numpyro/distributions/transforms.py [hunk 1]
```diff
@@ -1121,6 +1121,19 @@ def _transform_to_interval(constraint):
     )
 
 
+@biject_to.register(constraints.open_interval)
```
Comment (fehiepsi) [inline numpyro/distributions/transforms.py:1124]: You can add this to the above `def _transform_to_interval(constraint):`, so no need to reimplement the transform.

**Review by fehiepsi — APPROVED** [review]
Nice solution! Thanks, @MarcoGorelli!

**Comment by MarcoGorelli** [comment]
Thanks @fehiepsi !

I tried chaining the decorators, i.e.
```python
@biject_to.register( constraints.open_interval)
@biject_to.register(constraints.interval)
```
but that didn't work, so I modified `biject_to.register` to accept a list of constraints

File: numpyro/distributions/transforms.py [hunk 2]
```diff
@@ -1044,14 +1044,14 @@ class ConstraintRegistry(object):
...
+            return lambda factory: self.register(constraint_list, factory)
 
-        if isinstance(constraint, constraints.Constraint):
-            constraint = type(constraint)
-
-        self._registry[constraint] = factory
+        for constraint in constraint_list:
+            if isinstance(constraint, constraints.Constraint):
+                constraint = type(constraint)
+            self._registry[constraint] = factory
```
Comment (fehiepsi) [inline numpyro/distributions/transforms.py:1054]: Oops, really sorry, I think we need to return `factory` here. Could you revert to the previous commit and add
```
def register(...):
    ...
    return factory
```

Comment (MarcoGorelli) [reply numpyro/distributions/transforms.py]: Ah, right - that works, thanks!

**Review by fehiepsi — APPROVED** [review]
Thanks, @MarcoGorelli!

**Pull request merged**